sdk/core/Azure.Core/src/RequestContext.cs

```diff
@@ -44,8 +54,36 @@ public RequestContext()
         /// <param name="position"></param>
         public void AddPolicy(HttpPipelinePolicy policy, HttpPipelinePosition position)
         {
-            Policies ??= new();
-            Policies.Add((position, policy));
+            Policies ??= new Memory<HttpPipelinePolicy>(new HttpPipelinePolicy[PolicySections * SectionSize]);
+
+            switch (position)
+            {
+                case HttpPipelinePosition.PerCall:
+                    CheckPolicyCounter(PerCallPolicies);
+                    Policies.Value.Span[PerCallPolicies++] = policy;
+                    break;
+
+                case HttpPipelinePosition.PerRetry:
+                    CheckPolicyCounter(PerRetryPolicies);
+                    Policies.Value.Span[SectionSize + PerRetryPolicies++] = policy;
+                    break;
+
+                case HttpPipelinePosition.BeforeTransport:
+                    CheckPolicyCounter(BeforeTransportPolicies);
+                    Policies.Value.Span[2 * SectionSize + BeforeTransportPolicies++] = policy;
+                    break;
+
+                default:
+                    break;
+            }
+        }
+
+        private static void CheckPolicyCounter(int length)
+        {
+            if (length >= SectionSize)
+            {
+                throw new InvalidOperationException($"Cannot add more than {SectionSize} policies at a pipeline position.");
```

Comment:
This seems like an arbitrary limitation.

Reply:
Gone.